worlds/kh2/Rules.py

```diff
@@ -1,96 +1,1839 @@
+
+    def oc_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.BattlefieldsofWar, self.player, Amount)
+
+    def twtnw_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.WaytotheDawn, self.player, Amount)
+
+    def ht_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.BoneFist, self.player, Amount)
+
+    def tt_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.IceCream, self.player, Amount)
+
+    def pr_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.SkillandCrossbones, self.player, Amount)
+
+    def sp_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.IdentityDisk, self.player, Amount)
+
+    def stt_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.NamineSketches, self.player, Amount)
+
+    def dc_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.CastleKey, self.player, Amount)  # Using Dummy 13 for this
+
+    def hb_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.MembershipCard, self.player, Amount)
+
+    def pl_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.ProudFang, self.player, Amount)
+
+    def ag_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.Scimitar, self.player, Amount)
+
+    def bc_unlocked(self, state: CollectionState, Amount) -> bool:
+        return state.has(ItemName.BeastsClaw, self.player, Amount)
+
+    def at_three_unlocked(self, state: CollectionState) -> bool:
+        return state.has(ItemName.MagnetElement, self.player, 2)
+
+    def at_four_unlocked(self, state: CollectionState) -> bool:
+        return state.has(ItemName.ThunderElement, self.player, 3)
+
+    def hundred_acre_unlocked(self, state: CollectionState, amount) -> bool:
+        return state.has(ItemName.TornPages, self.player, amount)
+
+    def final_form_region_access(self, state: CollectionState) -> bool:
+        """
+        Can reach one of TT3,Twtnw post Roxas, BC2, LoD2 or PR2
+        """
+        # tt3 start, can beat roxas, can beat gr2, can beat xaldin, can beat storm rider.
+        final_leveling_access = [LocationName.MemorysSkyscaperMythrilCrystal, LocationName.GrimReaper2,
+                                 LocationName.Xaldin, LocationName.StormRider, LocationName.SunsetTerraceAbilityRing]
+        return any(
+                self.world.multiworld.get_location(location, self.player).can_reach(state) for location in
+                final_leveling_access
+        )
+
+    def limit_form_region_access(self, state: CollectionState) -> bool:
+        """
+        multi_form_region_access + namine sketches for leveling
+        """
+        multi_form_region_access = {
```

Comment:
these sets should also be module level instead of creating them every single time this rule gets called